BUG: r.Cancel() crashing

Good luck catching that in a unit test.  8-)

https://github.com/m-lab/ndt-server/blob/df9eced207b02932d82fd2bb4c2a7c1d66266054/legacy/s2c/s2c.go#L72

2019/03/19 18:41:32 ERROR: manageC2sTest context deadline exceeded
2019/03/19 18:41:32 S2C: Serving for test on [::]:45175
2019/03/19 18:41:44 S2C: Context Done! context deadline exceeded
2019/03/19 18:41:44 S2C: ctx done!
2019/03/19 18:41:44 Closing Test Responder
2019/03/19 18:41:44 Summarized data: &{0 103 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0}
2019/03/19 18:41:44 ERROR: manageS2cTest context canceled
2019/03/19 18:41:44 NDT: uploaded at 524.1984 and downloaded at 0.0000
35.225.75.192 - - [19/Mar/2019:18:41:15 +0000] "GET /ndt_protocol HTTP/1.1" 200 0
2019/03/19 18:41:44 ERROR S2C: sending message write tcp 4.35.94.50:43265->35.225.75.192:52660: write: broken pipe
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x867ffa]

goroutine 404 [running]:
github.com/m-lab/ndt-server/legacy/s2c.(*Responder).performTest.func1(0xa41b60, 0xc0002043b0, 0xc000208550, 0xc0008f6000, 0x2000, 0x2000, 0xc000192fc0)
	/go/src/github.com/m-lab/ndt-server/legacy/s2c/s2c.go:72 +0x18a
created by github.com/m-lab/ndt-server/legacy/s2c.(*Responder).performTest
	/go/src/github.com/m-lab/ndt-server/legacy/s2c/s2c.go:65 +0x148

--- a/legacy/s2c/s2c.go
+++ b/legacy/s2c/s2c.go
@@ -62,7 +62,7 @@ func (r *Responder) performTest(ws protocol.MeasuredConnection) {
 	// Create ticker to enforce timeout on
 	done := make(chan *web100.Metrics)
 
-	go func(r *Responder, ws protocol.MeasuredConnection) {
+	go func(r *Responder, ws protocol.MeasuredConnection, done chan<- *web100.Metrics) {
 		ws.StartMeasuring(r.Ctx)
 		startTime := time.Now()
 		endTime := startTime.Add(10 * time.Second)
@@ -79,7 +79,7 @@ func (r *Responder) performTest(ws protocol.MeasuredConnection) {
 		}
 		info.BytesPerSecond = float64(totalBytes) / float64(time.Since(startTime)/time.Second)
 		done <- info
-	}(r, ws)
+	}(r, ws, done)
 
 	log.Println("S2C: Waiting for test to complete or timeout")
 	select {

--- a/legacy/s2c/s2c.go
+++ b/legacy/s2c/s2c.go
@@ -62,7 +62,7 @@ func (r *Responder) performTest(ws protocol.MeasuredConnection) {
 	// Create ticker to enforce timeout on
 	done := make(chan *web100.Metrics)
 
-	go func() {
+	go func(r *Responder, ws protocol.MeasuredConnection) {
 		ws.StartMeasuring(r.Ctx)
 		startTime := time.Now()
 		endTime := startTime.Add(10 * time.Second)
@@ -79,7 +79,7 @@ func (r *Responder) performTest(ws protocol.MeasuredConnection) {
 		}
 		info.BytesPerSecond = float64(totalBytes) / float64(time.Since(startTime)/time.Second)
 		done <- info
-	}()
+	}(r, ws)
 
 	log.Println("S2C: Waiting for test to complete or timeout")
 	select {

